### PR TITLE
Fixing SpaceTime attention block in siglip

### PR DESCRIPTION
**Pull request by akshay18iitg**
## What this does                                                                                                                                                                         
                
  Fix the SpaceTime SigLIP video encoder's attention factorization to match MEM Eq. 3 (Reading B), and rescale the temporal sinusoidal PE so it cleanly encodes up to 20 frames. (🐛 Bug 
  + ⚡️ Performance)                                                                                                                                                                      
   
  Reading-B factorization in SpaceTimeEncoderLayerWrapper.forward — replaces the previous "two sub-attentions stacked with shared weights" implementation:                               
  - One LayerNorm on h + e(t) instead of two.                                                                                                                                          
  - One Q/K/V projection instead of two; Q and K reused across temporal and spatial passes (just permuted into each SDPA layout); only V is replaced by V_temp for the spatial pass.     
  - Two SDPA calls — temporal (causal over T, per spatial patch) then spatial (bidirectional over N, per timestep).                                                                    
  - One out_proj, one residual on the original hidden_states (not h_pe).                                                                                                                 
  - Removed the _TemporalSelfAttention private helper and the SiglipAttention import.                                                                                                    
  - Under Reading B the T=1 block is naturally identity (single-key temporal SDPA returns V unchanged + e(t=0)=0), so the existing T=1 short-circuit becomes a perf/precision            
  optimization rather than a correctness requirement.                                                                                                                                    
                                                                                                                                                                                         
  Temporal PE supports T up to 20 without aliasing — _build_temporal_sinusoidal_pe:                                                                                                      
  - Bumped max_period default from 4.0 → 40.0. With T=20 the lowest-frequency sinusoid completes ≤ ½ a cycle over the time range, so every timestep gets a unique low-frequency encoding.
  - Added a guard: num_frames - 1 > max_period raises with a clear suggested override. Callers needing T > 20 pass a larger max_period explicitly (rule of thumb: 2 · (T-1)).            
                                                                                                                                                                                         
  Preserved invariants: state_dict keys identical to vanilla SiglipEncoderLayer (no new params); e(t=0) = 0; current frame at index T-1; pe[max_num_frames - T:] slicing equivalence;    
  suppress_spacetime_temporal context manager; padded-history masking.                                                                                                                   
                                                                                                                                                                                         
 ## How it was tested                                                                                                                                                                      
                                                                                                                                                                                       
  Added 23 new CPU tests in tests/policies/test_pi07_video_encoder_cpu.py:                                                                                                               
   
  - test_pe_is_zero_at_t_eq_0 — 21 parametrized cases (T ∈ {1,2,3,4,8,16,32} × embed_dim ∈ {16,64,1152}) pinning the MEM e(t=0)=0 boundary.                                              
  - test_pe_supports_20_frames_without_aliasing — every pair of distinct rows in a T=20 PE differs by > 1e-2 (would have failed pre-fix).                                              
  - test_max_period_too_small_raises — alias guard + clean override path.                                                                                                                
  - test_t_above_1_with_full_pad_matches_t1 — encoder output at T=4 with all-padded history is assert_close to T=1 forward.                                                              
  - TestReadingBMaskStructure (3 tests) — patches F.scaled_dot_product_attention and asserts the wrapper fires exactly two SDPAs per forward at T>1 with the right mask shape: temporal  
  causal, spatial bidirectional. Includes a behavioral causality check: modifying only the current frame's input leaves the past-frame wrapper output unchanged.                         
                                                                                                                                                                                         
  All 65 CPU tests pass:                                                                                                                                                                 
  pytest tests/policies/test_pi07_video_encoder_cpu.py    # 65 passed in 255s  

## Checklist

- [X] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [X] My PR includes policy-related changes.
  - [X] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### Note: Before submitting this PR, please read the [contributor guideline](https://github.com/TensorAuto/OpenTau/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: src/opentau/policies/pi07/video_encoder.py [hunk 1]
```diff
@@ -365,31 +320,111 @@ def forward(
...
-        h_after_t = rearrange(t_res, "(b n) t d -> (b t) n d", n=n)
+        # ONE LayerNorm. Its output feeds BOTH attention passes.
+        z = self.layer_norm1(x_pe)
+
+        # ONE QKV projection. Q/K/V come from a single application of the
+        # layer's W_Q/W_K/W_V to LN1(h_pe). They are reshaped (not
+        # re-projected) into the temporal and spatial layouts below.
+        q = attn.q_proj(z).view(b, t, n, num_heads, head_dim)
+        k = attn.k_proj(z).view(b, t, n, num_heads, head_dim)
+        v = attn.v_proj(z).view(b, t, n, num_heads, head_dim)
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/policies/pi07/video_encoder.py]: Agreed.

File: src/opentau/policies/pi07/video_encoder.py [hunk 2]
```diff
@@ -365,31 +320,111 @@ def forward(
...
+                v_t,
+                attn_mask=None,
+                is_causal=True,
+                dropout_p=0.0,
+                scale=attn.scale,
+            )
+        # v_temp: (B*N, H, T, Dh). Permute back into (B, T, N, H, Dh) layout
+        # so it can feed the spatial pass as the V input. The .reshape at
+        # the end materializes the new strides into a contiguous tensor.
+        v_temp = v_temp.view(b, n, num_heads, t, head_dim).permute(0, 3, 1, 2, 4)
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/policies/pi07/video_encoder.py]: Agreed

**Comment by akshay18iitg** [comment]
@claude fix on agreed points

**Comment by shuheng-liu** [comment]
@claude fix 

Do we have unit tests that verifies the SpaceTime SigLIP video encoder gives the same output as a SigLIP vision encoder when you only pass in the current frame? If so, point it out in the comment. Otherwise, implement that test and make sure it passes.

**Review by WilliamYue37 — DISMISSED** [review]
I am fine with these changes as long as @shuheng-liu pytest request is resolved.

**Comment by akshay18iitg** [comment]
> @claude fix
> 
> Do we have unit tests that verifies the SpaceTime SigLIP video encoder gives the same output as a SigLIP vision encoder when you only pass in the current frame? If so, point it out in the comment. Otherwise, implement that test and make sure it passes.

Yes. we have the test

**Review by shuheng-liu — APPROVED** [review]

**Review by WilliamYue37 — APPROVED** [review]
